[Exclusivity] Handle mayRelease instructions conservatively in AccessnforcementOpts and LICM

shajrawi opened the pull request:
radar rdar://problem/42841839

Commit:
[Exclusivity] Handle mayRelease instructions conservatively in AccessEnforcementOpts and LICM

jrose-apple commented on lib/SILOptimizer/LoopTransforms/LICM.cpp:
```diff
@@ -534,9 +534,21 @@ static bool handledEndAccesses(BeginAccessInst *BI, SILLoop *Loop) {
   return true;
 }
 
+static bool isCoveredByScop(BeginAccessInst *BI, DominanceInfo *DT,
```
Is "scop" a typo or a term I don't know?

shajrawi replied:
Whoops! typo :) should be scope of course - thanks!

shajrawi commented:
@swift-ci Please test

Commit:
AccessEnforcementOpts: Refactoring
1) use vectors instead of MapVector for in/out of scope local maps
2) record nested conflict without creating a temporary copy

atrick reviewed (changes requested):
Your mayRelease changes look good.

Two somewhat separate areas for improvement in the existing code:

1. Use MapVector::remove_if instead of erase. Then remove all extra `while (changed)` loops, which are quadratic and complicate the code. They are in `visitSetForConflicts`, `detectApplyConflicts`, and `visitBeginAccess`. Also, remove the temporary map and extra loop in `mergeAccessStruct`.

2. Remove the temporary map and extra loop in `recordConflict`.

shajrawi commented:
@swift-ci Please test

shajrawi commented:
@swift-ci Please test

shajrawi commented:
@swift-ci Please test

shajrawi commented:
I changed the code per our offline discussion - we no longer use `MapVector` - added the changes as a separate commit in this PR

atrick commented on lib/SILOptimizer/Transforms/AccessEnforcementOpts.cpp:
```diff
@@ -142,13 +142,13 @@ namespace {
 /// Reachability results are stored here because very few accesses are
 /// typically in-progress at a particular program point,
 /// particularly at block boundaries.
-using DenseAccessMap = llvm::SmallMapVector<unsigned, BeginAccessInst *, 4>;
```
I think it would make more sense to call this `DenseAccessSet` (the vector part is an implementation detail of the Set).

shajrawi replied:
fixed in new PR

atrick commented on lib/SILOptimizer/Transforms/AccessEnforcementOpts.cpp:
```diff
@@ -398,16 +398,19 @@ void AccessConflictAndMergeAnalysis::addOutOfScopeAccess(
-  auto index = it->first;
-  info.outOfScopeConflictFreeAccesses.conflictFreeAccesses.erase(index);
+  auto *otherBegin = *it;
+  auto rmIt = std::find(
```
Why is this now using `find` instead of `conflictFreeAccesses.erase(otherBegin)`?

shajrawi replied:
You can't use `erase` on a non-iterator type. `it` is a reverse iterator.

atrick commented on lib/SILOptimizer/Transforms/AccessEnforcementOpts.cpp:
```diff
@@ -433,37 +436,26 @@ void AccessConflictAndMergeAnalysis::addOutOfScopeAccess(
-    accessStruct.conflictFreeAccesses.erase(index);
-  }
+  auto pred = [&](BeginAccessInst *it) {
+    auto rhsIt = std::find(RHSAccessStruct.conflictFreeAccesses.begin(),
```
Checking Set membership is done with `count`.

shajrawi replied:
Fixed - forgot to change that

atrick commented on lib/SILOptimizer/Transforms/AccessEnforcementOpts.cpp:
```diff
@@ -317,48 +319,55 @@ class AccessConflictAndMergeAnalysis {
-  info.inScopeConflictFreeAccesses.conflictFreeAccesses.insert(
-      std::make_pair(index, beginAccess));
+  assert(
+      std::find(info.inScopeConflictFreeAccesses.conflictFreeAccesses.begin(),
```
Set membership is done with `count`.

shajrawi replied:
Fixed - forgot to change that

atrick commented on lib/SILOptimizer/Transforms/AccessEnforcementOpts.cpp:
```diff
@@ -317,48 +319,55 @@ class AccessConflictAndMergeAnalysis {
+      auto &ai = result.getAccessInfo(*it);
+      ai.setSeenNestedConflict();
+    }
     accessStruct.conflictFreeAccesses.erase(it);
```
This still invalidates the iterator.

shajrawi replied:
Hmm yes: it might invalidate the iterator in `recordConflict`'s caller. not `it` in this function of course. I reverted to the old behavior in the caller

atrick reviewed (commented):
This doesn't look correct.

shajrawi commented:
Fixed https://github.com/apple/swift/pull/19140
